### PR TITLE
find_type fix

### PR DESCRIPTION
**Pull request by darkushin**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/28065
fixes: https://github.com/demisto/etc/issues/28337
fixes: https://github.com/demisto/etc/issues/28341
fixes: https://github.com/demisto/etc/issues/28359

## Description
Fixing an issue in find_type function, where the function doesn't recognize the type correctly.

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Review by yaakovi — CHANGES_REQUESTED** [review]
the whole purpose of `find_type` is to find the type of the file based on the *content* of the file and not his path.
the command `demisto-sdk validate -i ~/CortexXDR.yml` should figure the correct type and run the proper validations.
same for `demisto-sdk validate -i C:\CortexXDR.yml`

the example is for `validate` but relevant for all commands.

File: demisto_sdk/commands/common/tools.py [hunk 1]
```diff
@@ -769,59 +769,60 @@ def find_type(path: str = '', _dict=None, file_type: Optional[str] = None, ignor
...
-        if 'transformer' in _dict and 'keyTypeMap' in _dict:
-            return FileType.CLASSIFIER
+            if 'keyTypeMap' in _dict and 'transformer' in _dict:
+                return FileType.CLASSIFIER
 
-        if 'canvasContextConnections' in _dict:
+        if CONNECTIONS_DIR in path:
             return FileType.CONNECTION
 
         if 'mapping' in _dict:
```
Comment (roysagi) [inline demisto_sdk/commands/common/tools.py:815]: I think mappers are also under the Classifiers directory

File: demisto_sdk/commands/common/tools.py [hunk 2]
```diff
@@ -769,59 +769,60 @@ def find_type(path: str = '', _dict=None, file_type: Optional[str] = None, ignor
...
+        if DASHBOARDS_DIR in path:
             return FileType.DASHBOARD
 
-        if 'group' in _dict and LAYOUT_CONTAINER_FIELDS.intersection(_dict):
-            return FileType.LAYOUTS_CONTAINER
+        if LAYOUTS_DIR in path:
+            if 'group' in _dict and LAYOUT_CONTAINER_FIELDS.intersection(_dict):
+                return FileType.LAYOUTS_CONTAINER
+
+            return FileType.LAYOUT
```
Comment (bakatzir) [inline demisto_sdk/commands/common/tools.py:825]: As yaakovi said, please do not use the abs paths as it should work from non content repos as well.
for deferring layouts pre and post 6.0 you can also look at https://github.com/demisto/demisto-sdk/pull/742

File: demisto_sdk/commands/format/update_generic_json.py [hunk 3]
```diff
@@ -67,8 +67,9 @@ def remove_null_fields(self):
             a = yaml.safe_load(file_obj)
         schema_fields = a.get('mapping').keys()
         for field in schema_fields:
-            # We want to keep 'false' and 0 values.
-            if field in self.data and self.data[field] in (None, '', [], {}):
+            # We want to keep 'false' and 0 values, and avoid removing fields that are required in the schema.
+            if field in self.data and self.data[field] in (None, '', [], {}) and \
+                    not a.get('mapping').get(field).get('required'):
```
Comment (yaakovi) [inline demisto_sdk/commands/format/update_generic_json.py:72]:
```suggestion
                    not a.get('mapping', {}).get(field, {}).get('required'):
```

**Review by yaakovi — APPROVED** [review]

File: demisto_sdk/commands/common/tools.py [hunk 4]
```diff
@@ -822,15 +822,15 @@ def find_type(path: str = '', _dict=None, file_type: Optional[str] = None, ignor
         if 'brandName' in _dict and 'transformer' in _dict:
             return FileType.OLD_CLASSIFIER
 
-        if 'transformer' in _dict and 'keyTypeMap' in _dict:
-            return FileType.CLASSIFIER
+        if ('transformer' in _dict and 'keyTypeMap' in _dict) or 'mapping' in _dict:
+            if _dict.get('type') == 'classification':
```
Comment (roysagi) [inline demisto_sdk/commands/common/tools.py:826]:
```suggestion
            if _dict.get('type') and _dict.get('type') == 'classification':
```

let's be defensive here

File: demisto_sdk/commands/format/update_generic_json.py [hunk 5]
```diff
@@ -67,8 +67,9 @@ def remove_null_fields(self):
             a = yaml.safe_load(file_obj)
```
Comment (roysagi) [inline demisto_sdk/commands/format/update_generic_json.py:67]:
```suggestion
            schema_data = yaml.safe_load(file_obj)
```


A good opportunity to change a bad name to a more useful one

File: demisto_sdk/commands/common/tools.py [hunk 6]
```diff
@@ -822,15 +822,15 @@ def find_type(path: str = '', _dict=None, file_type: Optional[str] = None, ignor
...
+                return FileType.CLASSIFIER
+            else:
+                return FileType.MAPPER
```
Comment (roysagi) [inline demisto_sdk/commands/common/tools.py:829]:
```suggestion
            elif _dict.get('type') and 'mapping' in _dict.get('type'):
                return FileType.MAPPER
            else:
               return None
```

might give better results - here is my thought:
 - if the type is `classification` it is a classifier.
 - if the type has `mapping` in it - it is a mapper.
 - if not both then we cannot identify the file type and we should get the "file not identified error" hence return None. 
We can also print a warning message here to direct the user to the type field. but we need to make sure it looks good.

File: demisto_sdk/tests/integration_tests/validate_integration_test.py [hunk 7]
```diff
@@ -411,28 +411,28 @@ def test_missing_fromversion_field_in_new_classifier(self, mocker, repo):
...
-        new_classifier_copy = NEW_CLASSIFIER.copy()
-        new_classifier_copy['type'] = 'test'
-        classifier = pack.create_classifier('new_classifier', new_classifier_copy)
-        with ChangeCWD(pack.repo_path):
-            runner = CliRunner(mix_stderr=False)
-            result = runner.invoke(main, [VALIDATE_CMD, '-i', classifier.path], catch_exceptions=False)
-        assert f"Validating {classifier.path} as classifier" in result.stdout
-        assert 'Classifiers type must be classification' in result.stdout
-        assert result.exit_code == 1
+    # def test_invalid_type_in_new_classifier(self, mocker, repo):
```
Comment (roysagi) [inline demisto_sdk/tests/integration_tests/validate_integration_test.py:414]: we can either delete this or, by implementing my suggested changes make sure the non identified file error is raised.

**Review by roysagi — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/common/tools.py (hunk 4 above)
Comment (yaakovi) [reply demisto_sdk/commands/common/tools.py]: not really necessary as `None != 'classification'`

**Review by roysagi — APPROVED** [review]
looks good -

Maybe we can use this branch to check: https://github.com/demisto/etc/issues/28359 before merging? @bakatzir @yuvalbenshalom what do you think?

**Comment by darkushin** [comment]
> looks good -
> 
> Maybe we can use this branch to check: [demisto/etc#28359](https://github.com/demisto/etc/issues/28359) before merging? @bakatzir @yuvalbenshalom what do you think?

@roysagi I checked it for all file types (except `canvas` which I couldn't download). The validate command recognized the type correctly before and after formatting the file.

**Review by bakatzir — APPROVED** [review]
Well done.
please add a line or two in CHANGELOG.md

**Pull request merged**